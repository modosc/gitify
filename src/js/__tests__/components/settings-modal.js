import React from 'react'; // eslint-disable-line no-unused-vars
import { List, Map } from 'immutable';
import Enzyme, { shallow } from 'enzyme';
import Adapter from 'enzyme-adapter-react-16';
Enzyme.configure({
  adapter: new Adapter(),
});
import renderer from 'react-test-renderer';

const { ipcRenderer } = require('electron');

import {
  SettingsModal,
  mapStateToProps,
} from '../../components/settings-modal';

describe('components/settings-modal.js', () => {
  const props = {
    updateSetting: jest.fn(),
    fetchNotifications: jest.fn(),
    toggleSettingsModal: jest.fn(),
    logout: jest.fn(),
    isEitherLoggedIn: true,
    settings: Map({
      participating: false,
      playSound: true,
      showNotifications: true,
      markOnClick: false,
      openAtStartup: false,
      showSettingsModal: true,
      hasStarred: false,
      showAppIcon: 'both',
    }),
  };

  beforeEach(function() {
    ipcRenderer.send.mockReset();
    props.updateSetting.mockReset();
    props.toggleSettingsModal.mockReset();
  });

  it('should test the mapStateToProps method', () => {
    const state = {
      auth: Map({
        token: 123456,
        enterpriseAccounts: List(),
      }),
      settings: Map({
        participating: false,
        showSettingsModal: true,
      }),
    };

    const mappedProps = mapStateToProps(state);

    expect(mappedProps.isEitherLoggedIn).toBeTruthy();
    expect(mappedProps.settings.get('participating')).toBeFalsy();
    expect(mappedProps.settings.get('showSettingsModal')).toBeTruthy();
  });

  it('should close the modal on press of the escape key ', () => {
    spyOn(document, 'addEventListener').and.callFake((name, clb) => {
      if (name === 'keydown') {
        clb({ keyCode: 27 });
      }
    });

    spyOn(document, 'removeEventListener').and.callFake((name, clb) => {
      if (name === 'keydown') {
        clb({ keyCode: 27 });
      }
    });

    const wrapper = shallow(<SettingsModal {...props} />);
    expect(wrapper).toBeDefined();

    wrapper.instance().componentDidMount();
    expect(props.toggleSettingsModal).toHaveBeenCalledTimes(1);

    wrapper.instance().componentWillUnmount();
    expect(document.removeEventListener).toHaveBeenCalledTimes(1);
  });

  it('should render itself & its children (open modal)', () => {
    const tree = renderer.create(<SettingsModal {...props} />);
    expect(tree).toMatchSnapshot();
  });

  it('should render itself & its children (closed modal)', () => {
    const caseProps = {
      ...props,
      settings: props.settings.set('showSettingsModal', false),
    };

    const tree = renderer.create(<SettingsModal {...caseProps} />);
    expect(tree).toMatchSnapshot();
  });

  it('should redirect if logged out', () => {
    const caseProps = {
      ...props,
      isEitherLoggedIn: false,
    };

    const wrapper = shallow(<SettingsModal {...caseProps} />);
    expect(wrapper).toBeDefined();
    expect(wrapper.props().to).toEqual('/login');
  });

  it('should press the logout', () => {
    const wrapper = shallow(<SettingsModal {...props} />);
    expect(wrapper).toBeDefined();

    wrapper.find('.octicon-sign-out').parent().simulate('click');

    expect(props.logout).toHaveBeenCalledTimes(1);

    expect(ipcRenderer.send).toHaveBeenCalledTimes(1);
    expect(ipcRenderer.send).toHaveBeenCalledWith('update-icon');

    expect(props.toggleSettingsModal).toHaveBeenCalledTimes(1);
  });

  it('should call the componentWillReceiveProps method', function() {
    const wrapper = shallow(<SettingsModal {...props} />);
    expect(wrapper).toBeDefined();

    wrapper.setProps({
      ...props,
      settings: props.settings.set(
        'participating',
        !props.settings.get('participating')
      ),
    });
    expect(props.fetchNotifications).toHaveBeenCalledTimes(1);
  });

  it('should close the modal pressing the icon', () => {
    const wrapper = shallow(<SettingsModal {...props} />);
    expect(wrapper).toBeDefined();
    wrapper.find('.close').simulate('click');
    expect(props.toggleSettingsModal).toHaveBeenCalledTimes(1);
  });

  it('should toggle the showOnlyParticipating checbox', () => {
    const wrapper = shallow(<SettingsModal {...props} />);
    expect(wrapper).toBeDefined();

    wrapper
      .findWhere(node => node.props().label === 'Show only participating')
      .simulate('change', { target: { checked: true } });

    expect(props.updateSetting).toHaveBeenCalledTimes(1);
  });

  it('should toggle the playSound checbox', () => {
    const wrapper = shallow(<SettingsModal {...props} />);
    expect(wrapper).toBeDefined();

    wrapper
      .findWhere(node => node.props().label === 'Play sound')
      .simulate('change', { target: { checked: true } });

    expect(props.updateSetting).toHaveBeenCalledTimes(1);
  });

  it('should toggle the showNotifications checbox', () => {
    const wrapper = shallow(<SettingsModal {...props} />);
    expect(wrapper).toBeDefined();

    wrapper
      .findWhere(node => node.props().label === 'Show notifications')
      .simulate('change', { target: { checked: true } });

    expect(props.updateSetting).toHaveBeenCalledTimes(1);
  });

  it('should toggle the onClickMarkAsRead checbox', () => {
    const wrapper = shallow(<SettingsModal {...props} />);
    expect(wrapper).toBeDefined();

    wrapper
      .findWhere(node => node.props().label === 'On Click, Mark as Read')
      .simulate('change', { target: { checked: true } });

    expect(props.updateSetting).toHaveBeenCalledTimes(1);
  });

  it('should toggle the openAtStartup checbox', () => {
    const wrapper = shallow(<SettingsModal {...props} />);
    expect(wrapper).toBeDefined();

    wrapper
      .findWhere(node => node.props().label === 'Open at startup')
      .simulate('change', { target: { checked: true } });

    expect(props.updateSetting).toHaveBeenCalledTimes(1);
  });

  it('should toggle the showAppIcon radiogroup', () => {
    const wrapper = shallow(<SettingsModal {...props} />);
    expect(wrapper).toBeDefined();

    wrapper
      .find('RadioGroup')
      .simulate('change', { target: { value: 'both' } });

    expect(props.updateSetting).toHaveBeenCalledTimes(1);
    expect(props.updateSetting).toHaveBeenCalledWith('showAppIcon', 'both');
  });
});
