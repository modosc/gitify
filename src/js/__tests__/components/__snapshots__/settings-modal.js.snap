// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`components/settings-modal.js should render itself & its children (closed modal) 1`] = `null`;

exports[`components/settings-modal.js should render itself & its children (open modal) 1`] = `
<div>
  <div
    className="modal-content"
  >
    <div
      className="modal-header"
    >
      <h5
        className="modal-title"
      >
        Settings
      </h5>
      <button
        aria-label="Close"
        className="close"
        onClick={[Function]}
      >
        <span
          aria-hidden="true"
        >
          Ã—
        </span>
      </button>
    </div>
    <div
      className="modal-body"
    >
      <div
        className="row setting"
      >
        <div>
          <strong>
            Show only participating
          </strong>
          <input
            checked={false}
            onChange={[Function]}
            type="checkbox"
          />
        </div>
      </div>
      <div
        className="row setting"
      >
        <div>
          <strong>
            Play sound
          </strong>
          <input
            checked={true}
            onChange={[Function]}
            type="checkbox"
          />
        </div>
      </div>
      <div
        className="row setting"
      >
        <div>
          <strong>
            Show notifications
          </strong>
          <input
            checked={true}
            onChange={[Function]}
            type="checkbox"
          />
        </div>
      </div>
      <div
        className="row setting"
      >
        <div>
          <strong>
            On Click, Mark as Read
          </strong>
          <input
            checked={false}
            onChange={[Function]}
            type="checkbox"
          />
        </div>
      </div>
      <div
        className="row setting"
      >
        <div>
          <strong>
            Open at startup
          </strong>
          <input
            checked={false}
            onChange={[Function]}
            type="checkbox"
          />
        </div>
      </div>
      <div>
        <div>
          <strong>
            Both Icons
          </strong>
          <input
            type="radio"
            value="both"
          />
        </div>
        <div>
          <strong>
            Tray Icon
          </strong>
          <input
            type="radio"
            value="tray"
          />
        </div>
        <div>
          <strong>
            Dock Icon
          </strong>
          <input
            type="radio"
            value="dock"
          />
        </div>
      </div>
      <hr />
      Version: 
      0.0.1
    </div>
    <div
      className="modal-footer"
    >
      <button
        className="btn btn-outline-danger"
        onClick={[Function]}
      >
        <span
          className="octicon octicon-sign-out"
        />
         Logout from all accounts
      </button>
    </div>
  </div>
</div>
`;
